manager/dispatcher/dispatcher.go

```diff
@@ -1091,13 +1114,33 @@ func (d *Dispatcher) Assignments(r *api.AssignmentsRequest, stream api.Dispatche
 	}
 }
 
-func (d *Dispatcher) nodeRemove(id string, status api.NodeStatus) error {
+// markNodeNotReady sets the node state to some state other than READY
+func (d *Dispatcher) markNodeNotReady(id string, state api.NodeStatus_State, message string) error {
 	if err := d.isRunningLocked(); err != nil {
 		return err
 	}
 
+	// TODO(dperny) lock before or after we view the node?
 	d.nodeUpdatesLock.Lock()
-	d.nodeUpdates[id] = nodeUpdate{status: status.Copy(), description: d.nodeUpdates[id].description}
+
+	// lookup the node so we can get its old address
+	// TODO(dperny) is there a faster/better way to do this?
+	var node *api.Node
+	d.store.View(func(tx store.ReadTx) {
+		node = store.GetNode(tx, id)
+	})
+	if node == nil {
+		return ErrNodeNotFound
+	}
+
+	status := &api.NodeStatus{
+		State:   state,
+		Message: message,
+		Address: node.Status.Address,
+	}
+
+	// TODO(dperny) explain why we can just pluck the description out of nodeUpdates?
```

Comment:
Yeah, it's definitely a little confusing. The idea is that if a description update was previously "in flight" because `markNodeReady` put it in `d.nodeUpdates`, we don't want to accidentally overwrite it here. Let's say `markNodeReady` gets called, and then `markNodeNotReady` gets called right after, before the batch was processed. Without this weirdness, the description given to `markNodeReady` would be lost.